lib/schedule/data.ex

```diff
@@ -66,34 +80,87 @@ defmodule Schedule.Data do
-  def timepoints_on_route(
-        %__MODULE__{timepoints_by_route: timepoints_by_route},
-        route_id
-      ),
-      do: Map.get(timepoints_by_route, route_id, [])
+  @spec route_by_id(tables(), Route.id()) :: Route.t() | nil
+  def route_by_id(%{routes: routes_table}, route_id) do
+    case :mnesia.dirty_read(routes_table, route_id) do
+      [{_, _, route}] -> route
+      _ -> nil
+    end
+  end
 
-  @spec timepoint_names_by_id(t()) :: Timepoint.timepoint_names_by_id()
-  def timepoint_names_by_id(%__MODULE__{timepoint_names_by_id: timepoint_names_by_id}),
-    do: timepoint_names_by_id
+  @spec timepoints_on_route(tables(), Route.id()) :: [Timepoint.t()]
+  def timepoints_on_route(%{timepoints_by_route: timepoints_by_route_table}, route_id) do
+    timepoints_by_route_table
+    |> :mnesia.dirty_select([{{:_, route_id, :"$1"}, [], [:"$1"]}])
+    |> List.flatten()
+  end
 
-  @spec stop(t(), Stop.id()) :: Stop.t() | nil
-  def stop(%__MODULE__{stops: stops}, stop_id), do: stops[stop_id]
+  @spec timepoint_names_by_id(tables()) :: Timepoint.timepoint_names_by_id()
+  def timepoint_names_by_id(%{timepoint_names_by_id: timepoint_names_by_id_table}) do
+    timepoint_names_by_id_table
+    |> :mnesia.dirty_select([
+      {
+        {:_, :"$1", :"$2"},
+        [],
+        [[:"$1", :"$2"]]
+      }
+    ])
+    |> Map.new(fn [id, name] -> {id, name} end)
+  end
 
-  @spec trip(t(), Schedule.Trip.id()) :: Schedule.Trip.t() | nil
-  def trip(%__MODULE__{trips: trips}, trip_id), do: trips[trip_id]
+  @spec stop(tables(), Stop.id()) :: Stop.t() | nil
+  def stop(%{stops: stops_table}, stop_id) do
+    case :mnesia.dirty_read(stops_table, stop_id) do
+      [{^stops_table, _, stop}] -> stop
+      _ -> nil
+    end
+  end
 
-  @spec trips_by_id(t(), [Schedule.Trip.id()]) :: %{Schedule.Trip.id() => Schedule.Trip.t()}
-  def trips_by_id(%__MODULE__{trips: trips}, trip_ids) do
-    Map.take(trips, trip_ids)
+  @spec trip(tables(), Schedule.Trip.id()) :: Schedule.Trip.t() | nil
+  def trip(%{trips: trips_table}, trip_id) do
+    case :mnesia.dirty_read(trips_table, trip_id) do
+      [{^trips_table, _, _, trip}] -> trip
+      _ -> nil
+    end
   end
 
-  @spec block(t(), Block.id(), Service.id()) :: Block.t() | nil
-  def block(%__MODULE__{blocks: blocks}, schedule_id, block_id) do
-    Block.get(blocks, schedule_id, block_id)
+  @spec trips_by_id(tables(), [Schedule.Trip.id()]) :: %{Schedule.Trip.id() => Schedule.Trip.t()}
+  def trips_by_id(tables, trip_ids) do
+    selectors =
+      for trip_id <- trip_ids do
+        {{:_, trip_id, :_, :"$1"}, [], [:"$1"]}
+      end
+
+    tables.trips
+    |> :mnesia.dirty_select(selectors)
+    |> Map.new(fn trip -> {trip.id, trip} end)
+  end
+
+  @spec block(tables(), Hastus.Schedule.id(), Block.id()) :: Block.t() | nil
+  def block(%{blocks: blocks_table}, schedule_id, block_id) do
+    case :mnesia.dirty_read(blocks_table, {schedule_id, block_id}) do
+      [{_, _, _, block}] -> block
```

Comment:
**nitpick:** Unlike other similar uses of `dirty_read`, the `case` statement match doesn't pin / match on the table as the first element of the tuple. Probably wouldn't matter from a practical standpoint but it's probably nice to be consistent one way or the other.

Reply:
I'll go through and make these consistent.